anchor_optimization/optimize_anchors.py

```diff
@@ -0,0 +1,347 @@
+                base_dir = os.path.split(annotations)[0]
+                relative_path = row[0]
+                image_path = os.path.join(base_dir, relative_path)
+                img = Image.open(image_path)
+
+                if hasattr(img, "shape"):
+                    image_shape = img.shape
+                else:
+                    image_shape = (img.size[0], img.size[1], 3)
+
+                scale = compute_resize_scale(
+                    image_shape, min_side=image_min_side, max_side=image_max_side)
+                x1, y1, x2, y2 = list(map(lambda x: int(x) * scale, row[1:5]))
+
+            max_x = max(x2, max_x)
+            max_y = max(y2, max_y)
+
+            if include_stride:
+                entry = np.expand_dims(np.array([x1, y1, x2, y2]), axis=0)
+                entries = np.append(entries, entry, axis=0)
+            else:
+                width = x2 - x1
+                height = y2 - y1
+                entry = np.expand_dims(
+                    np.array([-width / 2, -height / 2, width / 2, height / 2]), axis=0)
+                entries = np.append(entries, entry, axis=0)
+
+    image_shape = [max_y, max_x]
+
+    if verbose:
+        print('Optimising anchors.')
+
+    bounds = []
+
+    for i in range(int((ratios - 1) / 2)):
+        bounds.append((1, 4))
+
+    for i in range(scales):
+        bounds.append((0.4, 2))
+
+    update_state = None
+    if threads == 1:
+        update_state = state
+
+    ARGS = (entries,
+            image_shape,
+            objective,
+            ratios,
+            include_stride,
+            SIZES,
+            STRIDES,
+            verbose,
+            update_state,
+            # return a single value ('to_tuple' parameter is set to False)
+            False,
+            threads)
+
+    result = scipy.optimize.differential_evolution(func=average_overlap,
+                                                   args=ARGS,
+                                                   # pass the *args as a tuple
+                                                   # [ SEE:
+                                                   # https://stackoverflow.com/q/32302654
+                                                   # ]
+                                                   mutation=mutation,
+                                                   updating=updating,
+                                                   workers=threads,
+                                                   bounds=bounds,
+                                                   popsize=popsize,
+                                                   seed=seed)
+
+    # ------------------------------------------------------------------------------------------------------ using the 'lambda' function gives an error because 'lambdas' are not pickleable and the 'workers' parameter requires that 'func' is pickleable [SEE: https://stackoverflow.com/questions/25348532/can-python-pickle-lambda-functions ]
```

Comment:
I think this comment could be removed, it takes up a lot of space but doesn't really add much value.

Reply:
I'll do this.